### ISSUE TITLE
[struct_pack] Compile error when derived class contains std::unique_ptr<derived> as a member variable.

### ISSUE THREAD

**Issue by WeaveAche** [alibaba/yalantinglibs#627]
### Search before asking

- [x] I searched the [issues](https://github.com/alibaba/yalantinglibs/issues) and found no similar issues.

### What happened + What you expected to happen
When working with derived classes in struct_pack, if a derived class contains a unique_ptr to any of the derived classes, there's a compilation error. It works fine in any other case (for e.g. in the code below, changing child to std::unique_ptr<base> works). Expected to work normally.

### Reproduction way
```
#include <bits/stdc++.h>
#include "ylt/struct_pack.hpp"

struct base {
  virtual uint32_t get_struct_pack_id() const = 0;
  virtual std::string get_name() const = 0;
  static std::unique_ptr<base> deserialize(std::string &serialized);
};

struct derived1 : public base {
  int b;
  virtual uint32_t get_struct_pack_id() const;
  std::string get_name() const override {
    return "derived1";
  }
};
STRUCT_PACK_REFL(derived1, b);

struct derived2 : public base {
  std::string c;
  std::unique_ptr<derived1> child;
  virtual uint32_t get_struct_pack_id() const;
  std::string get_name() const override { 
      return "derived2";
  }
};
STRUCT_PACK_REFL(derived2, c, child);

STRUCT_PACK_DERIVED_DECL(base, derived1, derived2);
STRUCT_PACK_DERIVED_IMPL(base, derived1, derived2);

std::unique_ptr<base> base::deserialize(std::string &serialized) {
    return struct_pack::deserialize_derived_class<base, derived1, derived2>(serialized).value();
}

int main() {
    auto d1 = derived1{};
    auto serialized = struct_pack::serialize<std::string>(d1);

    auto x = base::deserialize(serialized);
    std::cout << x->get_name() << std::endl;
}

```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by WeaveAche** [comment]
I think this is because `struct_pack::detail::is_base_class<derived1>` returns `true` even though it is not the base class.

**Comment by poor-circle** [comment]
Thanks for report, i will fix it later.

**Comment by poor-circle** [comment]
You can try to pull the fix code from #634 and verify it.

**Comment by poor-circle** [comment]
By the way, your code can be simply like this:
```cpp
#include <bits/stdc++.h>
#include "ylt/struct_pack.hpp"

struct base {
  virtual std::string get_name() const = 0;
  static std::unique_ptr<base> deserialize(std::string &serialized);
};

struct derived1 : public base {
  int b;
  std::string get_name() const override {
    return "derived1";
  }
};
STRUCT_PACK_REFL(derived1, b);

struct derived2 : public base {
  std::string c;
  std::unique_ptr<derived1> child;
  std::string get_name() const override { 
      return "derived2";
  }
};
STRUCT_PACK_REFL(derived2, c, child);

std::unique_ptr<base> base::deserialize(std::string &serialized) {
    return struct_pack::deserialize_derived_class<base, derived1, derived2>(serialized).value();
}

int main() {
    auto d1 = derived1{};
    auto serialized = struct_pack::serialize<std::string>(d1);

    auto x = base::deserialize(serialized);
    std::cout << x->get_name() << std::endl;
}

```
The `get_struct_pack_id()` & `STRUCT_PACK_DERIVED_IMPL` only needed when you want to serialized/deserialize std::unique_ptr<Base> by struct_pack::serialize/deserialize. struct_pack::deserialize_derived_class dont need this.
